embedded/appendable/remoteapp/remote_app.go

```diff
@@ -187,58 +179,47 @@ func (r *RemoteStorageAppendable) uploadChunk(chunkID int64, dontRemoveFile bool
 		defer metricsUploadFinished.Inc()
 
 		var newApp appendable.Appendable
-		cp := r.chunkedProcess(ctx)
+		cp := chain.New().WithContext(ctx).WithBackoff(r.backoff)
 
 		// Chunk data was already flushed when changing the active appendable
 
 		// Update size stats after flush
-		cp.Step(func() error {
+		cp.Next(func() error {
 			fi, err := os.Stat(fileName)
 			if err == nil {
 				r.mutex.Lock()
 				r.chunkInfos[chunkID].storageSize = fi.Size()
 				r.mutex.Unlock()
 			}
 			return nil
-		})
-
-		// Upload the chunk
-		cp.RetryableStep(func(retries int, delay time.Duration) (bool, error) {
+		}).NextWithRetry(func() error { // Upload the chunk
 			defer prometheus.NewTimer(metricsUploadTime).ObserveDuration()
 			err := r.rStorage.Put(ctx, r.remotePath+appName, fileName)
 			if err == nil {
-				return false, nil
+				return nil
 			}
 
 			metricsUploadRetried.Inc()
-			return true, nil
-		})
+			return err
+		}).NextWithRetry(func() error { // Wait for the chunk to become ready
 
-		// Wait for the chunk to become ready
-		cp.RetryableStep(func(retries int, delay time.Duration) (bool, error) {
 			exists, err := r.rStorage.Exists(ctx, r.remotePath+appName)
 			if err == nil && exists {
-				return false, nil
+				return nil
 			}
 
 			metricsUploadRetried.Inc()
-			return true, nil
-		})
-
-		// Open new appendable from the remote storage
-		cp.RetryableStep(func(retries int, delay time.Duration) (bool, error) {
+			return err
```

Comment:
This will work incorrectly if `err == nil && !exists` - at that situation we want to retry too since there may be some asynchronous processing on the remote storage and we want to ensure that the new object is already visible.

Reply:
ack